File: shell/macos/Sources/HeadroomKit/Models/Diagnostics.swift
public struct Diagnostics: Decodable, Sendable, Hashable {
    public let appVersion: String
    public let os: String?
    public let desktop: String?
    public let uptimeSecs: Double
    public let transports: [String]
    public let logLevel: String
    public let logLevelSource: String
    public let logFile: String?
    public let text: String

    enum CodingKeys: String, CodingKey {
        case os, desktop, transports, text
        case appVersion = "app_version"
        case uptimeSecs = "uptime_secs"
        case logLevel = "log_level"
        case logLevelSource = "log_level_source"
        case logFile = "log_file"
    }
}

File: shell/gnome/src/diagnostics.js
import { integer, isObject, number, text } from './fields.js';

export class DiagnosticsError extends Error {}

function decode(json) {
    try {
        return JSON.parse(json);
    } catch (error) {
        throw new DiagnosticsError(`Unreadable diagnostics from the Headroom service: ${error.message}`);
    }
}

function parseProviderCount(raw) {
    return {
        provider: text(raw.provider),
        accounts: integer(raw.accounts) ?? 0,
        usageHomes: integer(raw.usage_homes) ?? 0,
    };
}

export function parseDiagnostics(json) {
    const raw = decode(json);
    const report = isObject(raw) ? text(raw.text) : null;
    if (report === null) throw new DiagnosticsError('Unexpected diagnostics from the Headroom service');
    return {
        appVersion: text(raw.app_version),
        os: text(raw.os),
        desktop: text(raw.desktop),
        uptimeSecs: number(raw.uptime_secs),
        transports: Array.isArray(raw.transports) ? raw.transports.filter(text) : [],
        logLevel: text(raw.log_level),
        logLevelSource: text(raw.log_level_source),
        logFile: text(raw.log_file),
        providers: (Array.isArray(raw.providers) ? raw.providers.filter(isObject) : []).map(parseProviderCount),
        text: report,
    };
}
